server/src/main/java/org/elasticsearch/client/node/NodeClient.java

```diff
@@ -29,30 +30,43 @@
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.node.NodeClosedException;
 import org.elasticsearch.tasks.Task;
 import org.elasticsearch.tasks.TaskListener;
 import org.elasticsearch.tasks.TaskManager;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.RemoteClusterService;
 
+import java.util.Iterator;
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Supplier;
 
 /**
  * Client that executes actions on the local node.
  */
 public class NodeClient extends AbstractClient {
 
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    /**
+     * Set of all listeners that were passed to {@link #doExecute} but were not yet resolved. Used to fail outstanding listeners when
+     * {@link #close} is invoked while requests are in-flight.
+     */
+    private final Set<ActionListener<? extends ActionResponse>> outstandingListeners = ConcurrentCollections.newConcurrentSet();
+
     @SuppressWarnings("rawtypes")
     private Map<ActionType, TransportAction> actions;
 
     private TaskManager taskManager;
 
     /**
-     * The id of the local {@link DiscoveryNode}. Useful for generating task ids from tasks returned by
+     * The local {@link DiscoveryNode}. Useful for generating task ids from tasks returned by
      * {@link #executeLocally(ActionType, ActionRequest, TaskListener)}.
      */
-    private Supplier<String> localNodeId;
+    private Supplier<DiscoveryNode> localNode;
```

Comment:
Moved this to the full `DiscoveryNode` to be able to throw `NodeClosedException` and be consistent with what `TransportService` throws on a closed node. (adjustments to other tests are to enable this to work properly)